### DISCUSSION TITLE
Fields with `readonly()` doesn't uses `fillUsing`

### DISCUSSION THREAD

**Discussion by anthony-brindley** [laravel/nova-issues#4006] [Q&A]
- Laravel Version: 9.8.1
- Nova Version: 4.2.1

### Description:
When you have a fillUsing method and a dependsOn method on the same field, the fillUsing doesn't populate an attribute of the model. Removing the dependsOn fixes the issue.

### Detailed steps to reproduce the issue on a fresh Nova installation:

- Create a Model with a resource. 
- Add a 2 Text Fields
- On the 2nd field add a fillUsing method. In my example I have simply dd'd something to demonstrate the method is being run
- On the 2nd field add a dependsOn method. In my example I have made the field readonly if the 1st field equals 0;
- try submitting the form
-  with both the dependsOn and fillUsing, the form will submit and bypass the fill using.
- remove the dependsOn
- try submitting the form
- the network tab will show the dd worked demonstrating that the fillUsing method was called

Screenshot of my code: https://www.dropbox.com/s/z90i1xju2pwwefy/phpstorm64_2022-04-15_10-46-07.png?dl=0

Video: https://www.dropbox.com/s/ufx4w4yjv17lyl0/chrome_2022-04-15_10-48-53.mp4?dl=0

**Comment by crynobone** [comment]
Updating `dependsOn()` doesn't trigger`fillUsing()`. That only happened during actual database save.

**Comment by anthony-brindley** [comment]
But fillUsing should be triggered when you submit the form, every time shouldn't it so you can correctly populate the database based on the value of the field? I don't understand why it would be one or the other? For instance, I am using dependsOn to populate the options available in a select field, which will update when the parent is changed. 

In my real use case instance (not the mock), the i am looking to set an attribute on my model so I can pick it up on my observer and trigger an action if it's not null. Because fillUsing is not being triggered everytime, I am unable to consistently react to form submission.

**Comment by anthony-brindley** [comment]
So my real use case is Parent Field (Manufacturer). When changed, the Child (CarModel) updates it options to make all CarModels stored in the system for that Manufacturer available. If the user wants to add a CarModel, they should select 'New Model' which is an extra option I added into the option array. When this CarModel field is set to 'New Model', I have a text box, which is set to dependOn the CarModel field being 'New Model' - when this is the case it becomes editable. 

When the form is then submitted, I am looking for the new_model attribute and I can take an action to persist the model_name and get an id and update the correct attribute on the formData before persisting the Car. Because the dependsOn is not allowing the fillUsing to work, this isn't working and I think it's a bug, rather than a feature request because you would expect the fillUsing method to be called 100% of the time, irrelevant of other methods on the field?

**Comment by anthony-brindley** [comment]
Oh, and just to clarify, I submitted the form on both occasions to trigger a database save in my video. FillUsing was not actioned when the dependsOn was.

**Comment by crynobone** [comment]
The `dependsOn` doesn't use `fillUsing`, even up editing no model being query and doing so will be a breaking change. 

`FormData` is used to behave as close as possible to JavaScript interaction without having to duplicate the logic both on PHP and JavaScript. 

So changing `FormData` to `Eloquent` is major breaking change, changing the value of `FormData` is also a breaking change.

**Reply by anthony-brindley** [reply]
Hello? Did you see my reply to your comment at all? Could do with getting this bug resolved if possible?

**Reply by crynobone** [reply]
> Hello? Did you see my reply to your comment at all?

I read it but I also know that both features don't cause this to happen.

**Comment by anthony-brindley** [comment]
Maybe I'm missing something here - Im not asking the dependsOn to use fillUsing :). They are two completely independent things. dependsOn allows us to trigger changes to field attributes whereas the fillUsing method helps us populate things how we want when we submit the form. The dependsOn feature works great. I'm just saying that when using it, form submission bypasses fillUsing and it shouldn't.

**Comment by crynobone** [comment]
Update the title, `readonly()` field has never apply `fillUsing()` even on Nova 3 (and below).

**Reply by anthony-brindley** [reply]
I don’t feel your update of the title is accurate as I’ve never mentioned any problem with readonly  and I also don’t feel that you are understanding the problem - despite it being clear and explained, in plenty of detail. Is there someone else that can be added to the discussion please?

**Reply by anthony-brindley** [reply]
Simply put, the field which has dependsOn applied to it, doesn’t set the model attribute when submitting the form. Removing the dependsOn fixes the issue. Therefore, there is definitely some kind of bug.

**Reply by crynobone** [reply]
My response is based on your video example, if you still insist there is a bug, submit a full reproducing repository such as https://github.com/nova-issues and clearly show the issue.

**Comment by crynobone** [comment]
![image](https://user-images.githubusercontent.com/172966/163942136-f14c6614-3712-48a2-b222-6f37b61385c7.png)

```php
    /**
     * Get the fields displayed by the resource.
     *
     * @param  \Laravel\Nova\Http\Requests\NovaRequest  $request
     * @return array
     */
    public function fields(NovaRequest $request)
    {
        return [
            ID::make()->sortable(),

            MultiSelect::make('Text Keys', 'content_a')->options([
                'one' => 'One',
                'two' => 'Two',
                'three' => 'Three',
            ])->fillUsing(function($request, $model, $attribute, $requestAttribute) {
                ray('calling fill using from "Text Keys"');

                if ($request->exists($requestAttribute)) {
                    $value = $request[$requestAttribute];

                    $model->{$attribute} = in_array($value, ['']) ? null : json_decode($value, true);
                }
            }),

            MultiSelect::make('Numeric Keys', 'content_b')->options([
                '1' => 'One',
                '2' => 'Two',
                '3' => 'Three',
            ])->fillUsing(function($request, $model, $attribute, $requestAttribute) {
                ray('calling fill using from "Numeric Keys"');

                if ($request->exists($requestAttribute)) {
                    $value = $request[$requestAttribute];

                    $model->{$attribute} = in_array($value, ['']) ? null : json_decode($value, true);
                }
            })->readonly(),
        ];
    }
```

Above code proved that `fillUsing()` isn't called when `readonly()` is applied.

**Reply by anthony-brindley** [reply]
What about if readonly is a conditional and is set to false?
